### PR TITLE
Partial trace instrumentation in web frontend

### PR DESCRIPTION
**Pull request by ahmetb**
This patch does not cover tracing for RPCs, just the web frontend
endpoints and sub-operations.

Looking for feedback on styling and usage patterns.

cc: @rakyll

### CODE REVIEW

File: web/main.go [hunk 1]
```diff
@@ -161,15 +180,34 @@ func logHandler(h http.HandlerFunc) http.HandlerFunc {
 	}
 }
 
+// traceHandler decorates HandlerFunc with context that contains the trace span
+// and automatically closes the span after the request is handled.
+func (s *server) traceHandler(h func(context.Context, http.ResponseWriter, *http.Request)) http.HandlerFunc {
```
Comment (rakyll) [inline web/main.go]: Is there a reason you don't use https://godoc.org/cloud.google.com/go/trace/traceutil#HTTPHandler?

Then you can use, r.Context() to access to the span.

File: web/main.go [hunk 2]
```diff
@@ -180,7 +218,7 @@ func (s *server) authUser(r *http.Request) (user *pb.User, errFunc httpErrorWrit
 		return nil, badRequest, errors.Wrap(err, "failed to decode cookie")
 	}
 
-	userResp, err := s.getUser(userID)
+	userResp, err := s.getUser(trace.NewContext(ctx, span.NewChild("retrieve_user")), userID)
```
Comment (rakyll) [inline web/main.go]: Delegate the creation of the new child to getUser, you should just pass the parent ctx.

Each handler/method should be responsible for creating its own span and finishing it.

File: web/main.go (hunk 1 above)
Comment (ahmetb) [reply web/main.go]: @rakyll primary reason is that I didn't realize this package existed at all.

secondary reason is that I actually tried to create this functionality myself as

    func (HandlerFunc)HandlerFunc

however for some reason when I did `req = req.WithContext(ctx); h(w,r)` myself I realized the `r.URL.Query` was disappearing on the shallow copy.

File: web/main.go (hunk 2 above)
Comment (ahmetb) [reply web/main.go]: This is a bit unexpected/unintuitive for me. I don't quite understand why the called method should be aware of the caller and the fact that caller must have passed a parent ctx that it should be deriving a new span from?

Comment (rakyll) [reply web/main.go]: The caller method is not aware of the caller but the calling context. In your current case, you are leaking that userRepos will need to mark a span to authUser handler. authUser shouldn't know about the span(s) getUser require.

We will document this ownership model as soon as APIs are more stable.

Comment (ahmetb) [reply web/main.go]: Actually the code got a ton more clearer when I stopped doing

    ctx := r.Context()
    span := trace.FromContext(ctx)
    childMethod(trace.NewContext(ctx, span.NewChild("foo"))

and just pass the ctx down to the child method. Thanks!

**Comment by ahmetb** [comment]
Thanks for reviewing, this was helpful. Just fixed some stuff. Feel free to take a look again.

**Pull request merged**